### PR TITLE
EOS-15391 Enable Authorization for copy-object.

### PR DESCRIPTION
**Pull request by abhilekh268**

### CODE REVIEW

File: server/s3_copy_object_action.cc [hunk 1]
```diff
@@ -372,6 +374,22 @@ void S3CopyObjectAction::set_authorization_meta() {
   s3_log(S3_LOG_DEBUG, request_id, "Entering\n");
   auth_client->set_acl_and_policy(bucket_metadata->get_encoded_bucket_acl(),
                                   bucket_metadata->get_policy_as_json());
+  request->set_action_str("PutObject");
   next();
   s3_log(S3_LOG_DEBUG, "", "Exiting\n");
 }
+
+void S3CopyObjectAction::set_source_bucket_authorization_metadata() {
+  s3_log(S3_LOG_DEBUG, request_id, "Entering\n");
```
Comment (amitwac2608) [inline server/s3_copy_object_action.cc:383]: S3_LOG_INFO !!

File: server/s3_auth_client.h [hunk 2]
```diff
@@ -358,6 +358,8 @@ class S3AuthClient {
  public:
   S3AuthClient(std::shared_ptr<RequestObject> req,
                bool skip_authorization = false);
+  bool set_get_method = false;
+  std::string clientabsoulte_uri = "";
```
Comment (amitwac2608) [inline server/s3_auth_client.h:362]: std::string constructor would set the clientabsolute_uri to empty string, no need to do it explicitly.

File: server/s3_copy_object_action.cc [hunk 3]
```diff
@@ -372,6 +374,22 @@ void S3CopyObjectAction::set_authorization_meta() {
   s3_log(S3_LOG_DEBUG, request_id, "Entering\n");
   auth_client->set_acl_and_policy(bucket_metadata->get_encoded_bucket_acl(),
                                   bucket_metadata->get_policy_as_json());
+  request->set_action_str("PutObject");
   next();
   s3_log(S3_LOG_DEBUG, "", "Exiting\n");
 }
+
+void S3CopyObjectAction::set_source_bucket_authorization_metadata() {
```
Comment (amitwac2608) [inline server/s3_copy_object_action.cc:382]: can you add UT for this new function ?

**Review by amitwac2608 — COMMENTED** [review]
Looks good, couple of comments.

File: server/s3_copy_object_action.cc [hunk 4]
```diff
@@ -513,3 +515,27 @@ bool S3CopyObjectAction::if_source_and_destination_same() {
...
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  auth_client->set_get_method = true;
+  std::string uri = "/" + source_bucket_name + "/" + source_object_name;
+  auth_client->clientabsoulte_uri = uri;
+  auth_client->set_acl_and_policy(
+      source_object_metadata->get_encoded_object_acl(),
+      source_bucket_metadata->get_policy_as_json());
+  request->set_action_str("GetObject");
+  auth_client->check_authorization(
```
Comment (bkirunge7) [inline server/s3_copy_object_action.cc:537]: can we separate this, set function should only set values and authorization need to happen in different function

File: server/s3_copy_object_action.cc [hunk 5]
```diff
@@ -64,6 +64,8 @@ S3CopyObjectAction::S3CopyObjectAction(
 void S3CopyObjectAction::setup_steps() {
   s3_log(S3_LOG_DEBUG, request_id, "Setting up the action\n");
   ACTION_TASK_ADD(S3CopyObjectAction::validate_copyobject_request, this);
+  ACTION_TASK_ADD(S3CopyObjectAction::set_source_bucket_authorization_metadata,
```
Comment (bkirunge7) [inline server/s3_copy_object_action.cc:67]: separate authorization step

**Review by bkirunge7 — APPROVED** [review]

File: st/clitests/awss3api_spec.py [hunk 6]
```diff
@@ -18,14 +18,16 @@
...
 from framework import Config
 from framework import S3PyCliTest
 from awss3api import AwsTest
 from s3client_config import S3ClientConfig
 from aclvalidation import AclTest
+from auth import AuthTest
 
 # Helps debugging
-# Config.log_enabled = True
+Config.log_enabled = True
```
Comment (amitwac2608) [inline st/clitests/awss3api_spec.py:30]: I guess you forgot to comment this line ?

File: st/clitests/awss3api_spec.py [hunk 7]
```diff
@@ -1011,3 +1029,159 @@ def delete_object_list_file(file_name):
...
+    .execute_test().command_is_successful()
+
+os.environ["AWS_ACCESS_KEY_ID"] = source_access_key_args['AccessKeyId']
+os.environ["AWS_SECRET_ACCESS_KEY"] = source_access_key_args['SecretAccessKey']
+
+AwsTest('Owner account can do put object acl').put_object_acl("source-bucket", "source-object", "grant-full-control" ,"id=" + cross_access_key_args['CanonicalId'] )\
+.execute_test().command_is_successful()
+os.environ["AWS_ACCESS_KEY_ID"] = cross_access_key_args['AccessKeyId']
+os.environ["AWS_SECRET_ACCESS_KEY"] = cross_access_key_args['SecretAccessKey']
+AwsTest('Aws can copy object to same bucket').copy_object("source-bucket/source-object", "target-bucket", "source-object")\
```
Comment (amitwac2608) [inline st/clitests/awss3api_spec.py:1133]: I think test name is wrong !!

copy-paste error !!!

File: st/clitests/awss3api_spec.py [hunk 8]
```diff
@@ -1011,3 +1029,159 @@ def delete_object_list_file(file_name):
...
+#--------------------------create target bucket----------------------
+os.environ["AWS_ACCESS_KEY_ID"] = destination_access_key_args['AccessKeyId']
+os.environ["AWS_SECRET_ACCESS_KEY"] = destination_access_key_args['SecretAccessKey']
+
+AwsTest('Aws can create bucket').create_bucket("target-bucket").execute_test().command_is_successful()
+
+#---------------------------copy-object------------------------------
+os.environ["AWS_ACCESS_KEY_ID"] = cross_access_key_args['AccessKeyId']
+os.environ["AWS_SECRET_ACCESS_KEY"] = cross_access_key_args['SecretAccessKey']
+AwsTest('Aws can copy object to same bucket').copy_object("source-bucket/source-object", "target-bucket", "source-object")\
```
Comment (amitwac2608) [inline st/clitests/awss3api_spec.py:1096]: test message should be "Aws cannot copy ...." ?

**Review by amitwac2608 — COMMENTED** [review]
Some minor comments.

**Review by amitwac2608 — APPROVED** [review]

**Comment by abhilekh268** [comment]
jenkins job - 
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/270/

**Pull request merged**